base/compiler/ssair/passes.jl

```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
+        return key
+    else
+        return nothing
+    end
+end
+function perform_symbolic_evaluation(stmt::PhiNode, ssa_to_ssa, blockidx)
+    length(stmt.values) == 0 && return nothing
+
+    no_of_edges = length(stmt.edges)
+    key = Vector{Any}(undef, no_of_edges*2 + 2)
+    copyto!(key, stmt.edges)
+    key[no_of_edges+1] = :phi
+    key[no_of_edges+2] = blockidx
+    copyto!(key, no_of_edges+3, stmt.values)
+
+    firstval = nothing
+    allthesame = true
+    deletions = 0
+    for j in eachindex(stmt.values)
+        !isassigned(stmt.values, j) && return nothing
+
+        val = stmt.values[j]
+        if val isa SSAValue && ssa_to_ssa[val.id] == 0
+            deleteat!(key, j-deletions)
+            deletions += 1
+            deleteat!(key, no_of_edges+2-deletions+j)
+            deletions += 1
+        else
+            if val isa SSAValue
+                key[no_of_edges+2-deletions+j] = ssa_to_ssa[val.id]
+            end
+            if firstval === nothing
+                firstval = val
+            else
+                allthesame &= val === firstval
+            end
+        end
+    end
+    return svec(key...)
+end
+
+"""
+    gnv!(ir::IRCode) -> newir::IRCode
+
+Global Value Numbering (GVN) pass.
+
+This is based on the NewGVN pass in LLVM.
+
+GVN partitions all the statements in the IR into congruence classes. All elements in a congruence class are guaranteed to be the same.
+
+It implements the RPO value numbering algorithm based on the paper "SCC based value numbering" by L. Taylor Simpson.
+The elimination step is based on the implemenation in LLVM's NewGVN pass.
+"""
+function gvn!(ir::IRCode)
+    changed = true
+    ssa_to_ssa = fill(0, length(ir.stmts.stmt)) # Map from ssa to ssa of equivalent value
+    # Value type is SSAValue in order to reuse cache from it being boxed in svec
+    val_to_ssa = IdDict{SimpleVector, SSAValue}() # Map from value of an expression to ssa with equivalent value
+    sizehint!(val_to_ssa, length(ir.stmts.stmt))
+
+    while changed
+        changed = false
+
+        # Reverse Post Order traversal of dominator tree
+        for (blockidx, block) in enumerate(ir.cfg.blocks), i in block.stmts
+            if !(ir.stmts.stmt[i] isa Expr) & !(ir.stmts.stmt[i] isa PhiNode)
+                ssa_to_ssa[i] = i
+                continue
+            end
+
+            stmt::Union{Expr, PhiNode} = ir.stmts.stmt[i]
+
+            # IR_FLAG_NOTHROW is necessary, can't exclude it otherwise inference tests start to fail for some reason
+            # tmerge_types_slow starts to return Any instead of the desired result.
+            total_flags = IR_FLAG_CONSISTENT | IR_FLAG_EFFECT_FREE | IR_FLAG_NOTHROW
+            if !(ir.stmts.flag[i] & total_flags == total_flags)
+                ssa_to_ssa[i] = i
+                continue
+            end
+
+            value = if stmt isa Expr
+                if stmt.head === :the_exception # Do not eliminate exceptions
```

Comment:
This should not be IR_FLAG_CONSISTENT. If it is, something earlier has a bug that should be fixed rather than special casing here.

Reply:
That's what I figured but wasn't sure. I'll try and come up with some MWE's then.

Reply:
This doesn't work for example 
```julia
function f()
    try
        throw(42)
    catch err
        err
    end
end
```
```julia
julia> ir = Base.code_ircode(f, ()) |> only |> first
1 1 ─ %1 = $(Expr(:enter, #3))                                                               │
  2 ─      Main.throw(42)::Union{}                                                           │
  └──      unreachable                                                                       │
  3 ┄      $(Expr(:leave, 1))                                                                │
  4 ─ %5 = $(Expr(:the_exception))::Any                                                      │
  │        $(Expr(:pop_exception, :(%1)))::Any                                               │
  └──      return %5                                                                         │
                                                                                              

julia> ir[SSAValue(5)][:flag]
0x00000170
```
The consistent flag is not set on `f` however, `Base.infer_effects(f) = (!c,+e,!n,+t,+s,+m,+u)`. I didn't encounter an example where the consistent flag wasn't set on a `:the_exception`.